include/seqan3/index/fm_index.hpp

```diff
@@ -0,0 +1,336 @@
+//
+// ============================================================================
+
+/*!\file
+ * \author Christopher Pockrandt <christopher.pockrandt AT fu-berlin.de>
+ * \brief Contains the unidirectional seqan3::fm_index.
+ */
+
+ #include <sdsl/suffix_trees.hpp>
+
+ #include <range/v3/view/reverse.hpp>
+ #include <range/v3/view/transform.hpp>
+
+#include <seqan3/index/concept.hpp>
+#include <seqan3/index/fm_index_iterator.hpp>
+#include <seqan3/core/metafunction/range.hpp>
+
+namespace seqan3
+{
+
+/*!\addtogroup index
+ * \{
+ */
+
+// TODO(h-2): what is ingroup for (vs. addtogroup)?
+/*!\brief The default FM Index Configuration.
+ * \ingroup fm_index
+ *
+ * \details
+ *
+ * ### Space consumption
+ *
+ * SAMPLING_RATE = 16
+ *
+ * Asymptotic space consumption: TODO
+ *
+ * ### Running time
+ *
+ * T_BACKWARD_SEARCH: O(log sigma)
+ *
+ */
+struct fm_index_default_traits
+{
+    //!\brief Type of the underlying SDSL index.
+    using sdsl_index_type = sdsl::csa_wt<
+        sdsl::wt_blcd<
+            sdsl::bit_vector,
+            sdsl::rank_support_v<>,
+            sdsl::select_support_scan<>,
+            sdsl::select_support_scan<0>
+        >,
+        16,
+        10000000,
+        sdsl::sa_order_sa_sampling<>,
+        sdsl::isa_sampling<>,
+        sdsl::byte_alphabet
+    >;
+};
+
+/*!\brief The SeqAn FM Index.
+ * \ingroup fm_index
+ * \implements seqan3::fm_index_concept
+ * \tparam text_t The type of the text to be indexed; must satisfy seqan3::forward_range_concept.
+ * \tparam fm_index_traits The traits determining the implementation of the underlying SDSL index;
+                           must satisfy seqan3::fm_index_traits_concept.
+ * \details
+ *
+ * The seqan3::fm_index is a fast and space-efficient string index to search strings and collections of strings.
+ *
+ * ### General information
+ *
+ * Here is a short example on how to build an index and search a pattern using an iterator. Please note that there
+ * is a very powerful search module with a high-level interface seqan3::TODO that encapsulates the use of iterators.
+ *
+ * ```cpp
+ * #include <vector>
+ * #include <iostream>
+ * #include <seqan3/index/all.hpp>
+ *
+ * using namespace seqan3;
+ * using namespace seqan3::literal;
+ *
+ * int main(int argc, char ** argv)
+ * {
+ *     std::vector<dna4> genome{"ATCGATCGAAGGCTAGCTAGCTAAGGGA"_dna4};
+ *     fm_index<std::vector<dna4>> index{text}; // build the index
+ *
+ *     auto it = index.root(); // create an iterator pointing to the root of a virtual suffix tree
+ *     it.down("AAGG"_dna4); // search
+ *     std::cout << "Number of hits: " << it.count() << '\n'; // outputs: 2
+ *     std::cout << "Positions in the genome: ";
+ *     for (auto const & pos : it.locate()); // outputs: 8, 22
+ *         std::cout << pos << ' ';
+ *     std::cout << '\n';
+ *
+ *     return 0;
+ * }
+ * ```
+ *
+ * Even though the FM index is originally a prefix tree and one performs backward searches, it is implemented as a
```

Comment:
[...] a prefix tree and ~~one~~ performs backward searches [...]

Reply:
Hm ... idk ... strictly speaking the index itself doesn't do backward searches but rather the iterator / the user / ...

Reply:
I changed it to "[...] is originally a prefix tree and uses backward searches [...]"